lxd/api_cluster.go

```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
+		// Get the node.
+		node, err = tx.GetNodeByName(nodeName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to get node by name")
+		}
+
+		if node.State == db.ClusterMemberStatePending {
+			return fmt.Errorf("Cannot evacuate or restore a pending node")
+		}
+
+		// Set node status to EVACUATED to prevent instances from being created.
+		err = tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateEvacuated)
+		if err != nil {
+			return errors.Wrap(err, "Failed to update node status")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Do nothing if the node is already evacuated.
+	if node.State == db.ClusterMemberStateEvacuated {
+		return response.SmartError(fmt.Errorf("Node is already evacuated"))
+	}
+
+	reverter := revert.New()
+	defer reverter.Fail()
+
+	// Ensure node is put into its previous state if anything fails.
+	reverter.Add(func() {
+		d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateCreated)
+
+			return nil
+		})
+	})
+
+	// The instances are retrieved in a separate transaction, after the node is in EVACUATED state.
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// If evacuating, consider only the instances on the node which needs to be evacuated.
+		dbInstances, err = tx.GetInstances(db.InstanceFilter{Node: nodeName})
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances := make([]instance.Instance, len(dbInstances))
+
+	for i, dbInst := range dbInstances {
+		inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+		if err != nil {
+			return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+		}
+
+		instances[i] = inst
+	}
+
+	var targetNodeName string
+	var targetNode db.NodeInfo
+
+	run := func(op *operations.Operation) error {
+		metadata := make(map[string]interface{})
+
+		// Stop all instances if needed
+		for _, inst := range instances {
+			if inst.IsRunning() {
+				metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+				op.UpdateMetadata(metadata)
+
+				// Be nice
+				timeout := inst.ExpandedConfig()["boot.host_shutdown_timeout"]
+				val, err := strconv.Atoi(timeout)
+				if err != nil {
+					val = 30
+				}
+
+				err = inst.Shutdown(time.Duration(val) * time.Second)
+				if err != nil {
+					return errors.Wrapf(err, "Failed to stop instance %q", inst.Name())
+				}
+
+				inst.VolatileSet(map[string]string{"volatile.last_state.power": "RUNNING"})
+			}
+
+			if !inst.IsMigratable() {
+				continue
+			}
+
+			metadata["evacuation_progress"] = fmt.Sprintf("Migrating %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			inst.VolatileSet(map[string]string{"volatile.evacuate.origin": nodeName})
+
+			startInstance := inst.LocalConfig()["volatile.last_state.power"] == "RUNNING"
```

Comment:
You could do a `isRunning := inst.IsRunning()` at the beginning of the loop, then use that both for the if statement and then later to know whether to start the instance back up.